xla/service/gpu/cancel_all_gather_dynamic_slice_test.cc

```diff
@@ -0,0 +1,237 @@
+                                               /*num_partitions=*/8,
+                                               /*expect_change=*/true));
+}
+
+TEST_F(CancelAllGatherDynamicSliceTest, AllPartitionsWithReshapeMultipleUsers) {
+   absl::string_view hlo_string = R"(
+HloModule AllGather
+ 
+ENTRY %AllGather {
+  %param = f32[32,8,128]{2,1,0} parameter(0)
+  %ag = f32[256,8,128]{2,1,0} all-gather(%param), replica_groups={{0,1,2,3,4,5,6,7}},
+    dimensions={0}, channel_id=1, use_global_device_ids=true
+  %reshape = f32[256,8,64,2]{3,2,1,0} reshape(%ag)
+  %pid = u32[] partition-id()
+  %pid_s32 = s32[] convert(%pid)
+  %slice_size = s32[] constant(32)
+  %offset = s32[] multiply(%pid_s32, %slice_size)
+  %zero = s32[] constant(0)
+  ROOT %ds = f32[32,8,64,2]{3,2,1,0} dynamic-slice(%reshape, %offset, %zero, %zero, %zero),
+    dynamic_slice_sizes={32,8,64,2}
+}
+)";
+  TF_ASSERT_OK_AND_ASSIGN(auto module, RunPass(hlo_string,
+                                               /*num_replicas=*/1,
+                                               /*num_partitions=*/8,
+                                               /*expect_change=*/true));
+}
+
+TEST_F(CancelAllGatherDynamicSliceTest, AllPartitionsWithReshapeOnSliceDim) {
+  absl::string_view hlo_string = R"(
+HloModule AllGather
+
+ENTRY %AllGather {
+  %param = f32[32,8,128]{2,1,0} parameter(0)
+  %ag = f32[256,8,128]{2,1,0} all-gather(%param), replica_groups={{0,1,2,3,4,5,6,7}},
+    dimensions={0}, channel_id=1, use_global_device_ids=true
+  %reshape = f32[2048,128]{1,0} reshape(%ag)
+  %pid = u32[] partition-id()
+  %pid_s32 = s32[] convert(%pid)
+  %slice_size = s32[] constant(256)
+  %offset = s32[] multiply(%pid_s32, %slice_size)
+  %zero = s32[] constant(0)
+  ROOT %ds = f32[256,128]{1,0} dynamic-slice(%reshape, %offset, %zero),
+    dynamic_slice_sizes={256,128}
+}
+)";
+  TF_ASSERT_OK_AND_ASSIGN(auto module, RunPass(hlo_string,
+                                               /*num_replicas=*/1,
+                                               /*num_partitions=*/8,
+                                               /*expect_change=*/false));
+}
+
+TEST_F(CancelAllGatherDynamicSliceTest, NoAllGather) {
+  absl::string_view hlo_string = R"(
+HloModule NoAllGather
+
+ENTRY %NoAllGather {
+  %param = f32[32,8,128]{2,1,0} parameter(0)
+  %pid = u32[] partition-id()
+  %pid_s32 = s32[] convert(%pid)
+  %slice_size = s32[] constant(32)
+  %offset = s32[] multiply(%pid_s32, %slice_size)
+  %zero = s32[] constant(0)
+  ROOT %ds = f32[32,8,128]{2,1,0} dynamic-slice(%param, %offset, %zero, %zero),
+    dynamic_slice_sizes={32,8,128}
+}
+)";
+  TF_ASSERT_OK_AND_ASSIGN(auto module, RunPass(hlo_string,
+                                               /*num_replicas=*/1,
+                                               /*num_partitions=*/1,
+                                               /*expect_change=*/false));
+  EXPECT_THAT(module->entry_computation()->root_instruction(),
```

Comment:
Can we consider a filecheck test for non-trivial patterns like this? 

all_reduce_splitter_test.cc#202 is an example of a good one